### PR TITLE
llvm18: new recipe

### PR DESCRIPTION
**Pull request by davidkaroly**
Successful build on x86_64 and the resulting toolchain can build simple C/C++ examples.
<s>Haven't tried it on 32-bit yet.</s>Successful build and smoke test on x86 secondary arch.

### CODE REVIEW

File: sys-devel/llvm/llvm18-18.1.1.recipe [hunk 1]
```diff
@@ -0,0 +1,788 @@
...
+	devel:libclangtoolinginclusionsstdlib$secondaryArchSuffix = $portVersion
+	devel:libclangtoolingrefactoring$secondaryArchSuffix = $portVersion
+	devel:libclangtoolingsyntax$secondaryArchSuffix = $portVersion
+	devel:libclangtransformer$secondaryArchSuffix = $portVersion
+	lib:libclang$secondaryArchSuffix = $portVersionCompat
+	lib:libclang_cpp$secondaryArchSuffix = $portVersionCompat
+	"
+REQUIRES_clang="
+	haiku$secondaryArchSuffix
+	lib:libLLVM$secondaryArchSuffix
```
Comment (korli) [inline sys-devel/llvm/llvm18-18.1.1.recipe:407]: since clang16 it looks like clang defaults to ld.ldd but the dependency is missing

**Comment by korli** [comment]
It would nice to drop $secondaryArchSuffix on every cmd: provides. It's very inconvenient to have to specify cmd:clang_x86

**Comment by davidkaroly** [comment]
> It would nice to drop $secondaryArchSuffix on every cmd: provides. It's very inconvenient to have to specify cmd:clang_x86

Not quite sure... the $secondaryArchSuffix'es were introduced in #9593 to fix a build issue for qt on 32-bit (see #9562)
This will need some more investigation.

**Comment by korli** [comment]
it seems clang since at least version 16 generates DT_RUNPATH instead of DT_RPATH. 
I had to use export LDFLAGS="-Wl,--disable-new-dtags" to have DT_RPATH generated.

**Comment by davidkaroly** [comment]
updated patch:
- uplift to llvm-18.1.2
- add lld to clang package's dependencies

**Comment by davidkaroly** [comment]
> it seems clang since at least version 16 generates DT_RUNPATH instead of DT_RPATH. I had to use export LDFLAGS="-Wl,--disable-new-dtags" to have DT_RPATH generated.

not sure about this one, I see some commits in haiku repo related to rpath and runpath, not sure if anything is needed in this recipe?

**Comment by korli** [comment]
> not sure about this one, I see some commits in haiku repo related to rpath and runpath, not sure if anything is needed in this recipe?

For r1beta4 some workaround would be needed.

**Comment by davidkaroly** [comment]
> > It would nice to drop $secondaryArchSuffix on every cmd: provides. It's very inconvenient to have to specify cmd:clang_x86
> 
> Not quite sure... the $secondaryArchSuffix'es were introduced in #9593 to fix a build issue for qt on 32-bit (see #9562) This will need some more investigation.

actually we can override `CMAKE_INSTALL_BINDIR`, so far the results are promising. the binaries are installed in `bin/` and this time the paths in `LLVMExports-release.cmake` also seem to line up correctly so we should be able to avoid the issue seen in #9562

**Comment by Begasus** [comment]
I haven't checked this yet, but pushed an update for clazy yesterday, on 64bit the build went fine but for 32bit it failed, maybe related to the comments by @korli ?
```
CMake Error at cmake/FindLLVM.cmake:125 (string):
 string sub-command REPLACE requires at least four arguments.
Call Stack (most recent call first):
 cmake/FindClang.cmake:46 (find_package)
 CMakeLists.txt:29 (find_package)
```
https://build.haiku-os.org/buildmaster/master/x86_gcc2/logviewer.html?buildruns/6154/builds/83288.log

**Comment by Begasus** [comment]
@davidkaroly maybe have a look at https://github.com/haikuports/haikuports/pull/10501 I had to revert the move for the binaries on 32bit?

**Comment by brad0** [comment]
> EDIT: a source I'm checking into requires LLVM18.1.4 minimum, maybe bump this one to 18.1.6 if possible?

18.1.7 was just released.

**Comment by brad0** [comment]
Also looking at the recipe the patch for clang/lib/Lex/InitHeaderSearch.cpp doesn't do anything anymore as the header path handling was moved to clang/lib/Driver/ToolChains/Haiku.cpp -- Haiku::AddClangSystemIncludeArgs(). This changed with 18.

**Comment by Begasus** [comment]
> > EDIT: a source I'm checking into requires LLVM18.1.4 minimum, maybe bump this one to 18.1.6 if possible?
> 
> 18.1.7 was just released.

Yeah saw that yesterday, only 1 listed the version at repology though (14 today), so I checked 18.1.6 first. :)

**Comment by Begasus** [comment]
> Also looking at the recipe the patch for clang/lib/Lex/InitHeaderSearch.cpp doesn't do anything anymore as the header path handling was moved to clang/lib/Driver/ToolChains/Haiku.cpp -- Haiku::AddClangSystemIncludeArgs(). This changed with 18.

Patchset still applies on 18.1.7, @korli, @davidkaroly or someone else, could one of you look into the above if it's still required? (ps, did build for kdevelop and latest qt-creator with 18.1.6, both went fine and no issues on runtime so far).

**Comment by korli** [comment]
> > Also looking at the recipe the patch for clang/lib/Lex/InitHeaderSearch.cpp doesn't do anything anymore as the header path handling was moved to clang/lib/Driver/ToolChains/Haiku.cpp -- Haiku::AddClangSystemIncludeArgs(). This changed with 18.
> 
> Patchset still applies on 18.1.7, @korli, @davidkaroly or someone else, could one of you look into the above if it's still required? (ps, did build for kdevelop and latest qt-creator with 18.1.6, both went fine and no issues on runtime so far).

It could mean that the header search is broken on x86 secondary arch.

**Comment by brad0** [comment]
I don't know how I missed the build issue with llvm-jitlink as I swear it went through a build without issue.

https://github.com/llvm/llvm-project/commit/2e0a105761cd33e42b915da5aab35c966daf8241

I pushed the llvm-jitlink fix.

https://github.com/llvm/llvm-project/commit/d3bcd9b16badac301d4f29cf401387ae5e14999f

**Comment by Begasus** [comment]
> It could mean that the header search is broken on x86 secondary arch.

Guess it's time to check this on 32bit :) 64bit went fine. @brad0 thanks for taking the time to look into this! +1

**Comment by Begasus** [comment]
Build completed on 32bit for LLVM18.1.7 based on this one, llvm-config gives correct paths.

**Comment by brad0** [comment]
Also looking at coming up with a slightly better llvm-config diff that doesn't use #ifdef.

Something like this..

```
diff --git a/llvm/tools/llvm-config/llvm-config.cpp b/llvm/tools/llvm-config/llvm-config.cpp
index d5b76b1bb6c1..30cb748b10b3 100644
--- a/llvm/tools/llvm-config/llvm-config.cpp
+++ b/llvm/tools/llvm-config/llvm-config.cpp
@@ -327,6 +327,7 @@ int main(int argc, char **argv) {
   std::string ActivePrefix, ActiveBinDir, ActiveIncludeDir, ActiveLibDir,
               ActiveCMakeDir;
   std::string ActiveIncludeOption;
+  const Triple HostTriple(Triple::normalize(LLVM_HOST_TRIPLE));
   if (IsInDevelopmentTree) {
     ActiveIncludeDir = std::string(LLVM_SRC_ROOT) + "/include";
     ActivePrefix = CurrentExecPrefix;
@@ -366,7 +367,10 @@ int main(int argc, char **argv) {
       sys::fs::make_absolute(ActivePrefix, Path);
       ActiveBinDir = std::string(Path);
     }
-    ActiveLibDir = ActivePrefix + "/lib" + LLVM_LIBDIR_SUFFIX;
+    if (HostTriple.isOSHaiku())
+      ActiveLibDir = ActivePrefix + "/develop/lib" + LLVM_LIBDIR_SUFFIX;
+    else
+      ActiveLibDir = ActivePrefix + "/lib" + LLVM_LIBDIR_SUFFIX;
     {
       SmallString<256> Path(LLVM_INSTALL_PACKAGE_DIR);
       sys::fs::make_absolute(ActivePrefix, Path);
@@ -385,7 +389,6 @@ int main(int argc, char **argv) {
   StringRef SharedExt, SharedVersionedExt, SharedDir, SharedPrefix, StaticExt,
       StaticPrefix, StaticDir = "lib";
   std::string DirSep = "/";
-  const Triple HostTriple(Triple::normalize(LLVM_HOST_TRIPLE));
   if (HostTriple.isOSWindows()) {
     SharedExt = "dll";
     SharedVersionedExt = LLVM_DYLIB_VERSION ".dll";
```

**Comment by Begasus** [comment]
> Also looking at coming up with a slightly better llvm-config diff that doesn't use #ifdef.

Should've know better I gues ;) replaced the above with the one currently in, didn't apply :)

File: sys-devel/llvm/llvm18-18.1.2.recipe [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+functions. The low level APIs are the unw_* functions which are an \
+interface defined by the old HP libunwind project."
+PROVIDES_libunwind_devel="
+	llvm18${secondaryArchSuffix}_libunwind_devel = $portVersion
+	devel:libunwind$secondaryArchSuffix = $libunwindLibVersion compat >= $libunwindSoVersion
+	"
+REQUIRES_libunwind_devel="
+	llvm18${secondaryArchSuffix}_libunwind == $portVersion base
+	haiku${secondaryArchSuffix}_devel
+	"
```
Comment (Begasus) [inline sys-devel/llvm/llvm18-18.1.2.recipe:514]: Shouldn't there be a CONFLICTS here? Looking at the contect of the libunwind packages for LLVM17 and LLVM18 I  see both have the same files?

llvm

**Comment by Begasus** [comment]
Pushed a commit to a branch on my fork: https://github.com/Begasus/haikuports/commit/57b1c111aacb8e3d9ad8b4620f8e662d43d18043

This contains the work from here and suggested changes. Could be nice if someone else beside me could check this. :)

**Comment by brad0** [comment]
> > Also looking at coming up with a slightly better llvm-config diff that doesn't use #ifdef.
> 
> Should've know better I gues ;) replaced the above with the one currently in, didn't apply :)

I should have clarified. That is against llvm main. I think the code is slightly different for 18.

**Comment by brad0** [comment]
> Pushed a commit to a branch on my fork: [Begasus@57b1c11](https://github.com/Begasus/haikuports/commit/57b1c111aacb8e3d9ad8b4620f8e662d43d18043)
> 
> This contains the work from here and suggested changes. Could be nice if someone else beside me could check this. :)

Also remove the clang/lib/Lex/InitHeaderSearch.cpp patch.

**Comment by Begasus** [comment]
> Also remove the clang/lib/Lex/InitHeaderSearch.cpp patch.

Re-launching new build without it. If things are still good I'll push it up to my branch to check. Thanks!

**Comment by Begasus** [comment]
OK, think I've done about all I can for this, latest changes at: https://github.com/haikuports/haikuports/compare/master...Begasus:haikuports:llvm18.1.7

On 64bit and 32bit, full build for LLVM18
On 64bit, checked build for Qt-Creator, KDevelop and Clazy, both Qt-Creator and Kdevelop show up the Clazy plugin and it's checks.
On 32bit, checked build for KDevelop and Clazy, Clazy plugin shows up correctly and checks are shown.

If no one objects I could push the changes into this PR so we could merge? (if nothing else is needed)

PS for the record, KDevelop uses the cmake config files from LLVM while Clazy and Qt-Creator use llvm-config from the package, so both detections work fine on both arch's.

**Comment by Begasus** [comment]
Thanks to all involved! 👍

**Comment by brad0** [comment]
@Begasus Just FYI there was a regression release of 18.1.8.

**Comment by Begasus** [comment]
@brad0 are we infected by any of those regressions mentioned at: https://discourse.llvm.org/t/18-1-8-released/79725#release-notes-2 ?

**Comment by Begasus** [comment]
Luckaly no changes needed for an update, build completed on 64bit just now. If things are fine on 32bit maybe we coculd update? pokes @korli @davidkaroly :)

**Pull request merged**